ALC Period Info Widget

Matt-Cumber opened the pull request:
**Description of work.**
Last release a new table widget was added to the muon interfaces which displays period information from the currently loaded run. This PR implements the same table for the ALC interface too. Since ALC is written in C++ and MA/FDA in python, the widget was rewritten in C++ along with it's tests and uses sip to export the widget for use in the python GUI's

**To test:**
Open ALC interface
Load HIFI 110542
Click the periods button
There should be 8 periods with a table containing data similar to this (taken from Wimda)
![image](https://user-images.githubusercontent.com/46685490/124495371-ee2eec80-ddaf-11eb-93b5-0805502abda1.png)
The choice of columns was already decided in the previous [PR](https://github.com/mantidproject/mantid/pull/30939)
The same table should be reproducible in Muon Analysis and Frequency Domain Analysis.
Go to the grouping tab in both interface and click the periods button
Do some stress testing such as clicking the button with no data, load new data with the table open, clear the MA/FDA Guis with it open etc.

Fixes #31798 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Matt-Cumber committed:
Re #31798 Set Up Period Info And Basic Display In C++

Matt-Cumber committed:
Re #31798 Period Info Widget Complete

Widget is complete in C++ and the ALC can now display period information in the same way as Muon Analysis

Matt-Cumber committed:
Re #31798 Set Up Unit Tests

Matt-Cumber committed:
Re #31798 Set Up Sip For Use Of Widget In Python

Matt-Cumber committed:
Re #31798 Unit Tests And Working Version In Muon Analysis

Matt-Cumber committed:
Re #31798 Unit Tests And Working Version In Muon Analysis

Matt-Cumber committed:
Re #31798 Clean Up Code

AnthonyLim23 commented on qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h:
```diff
@@ -0,0 +1,66 @@
+  void setWidgetTitleRuns(const std::string &title);
+  /// Get the title of the widget
+  std::string getWidgetTitleRuns() const;
+  /// Set the number of sequences that was gathered
```
what do you mean by sequence?

Matt-Cumber replied:
To be honest this is just the name given in the nexus file, I am not 100% sure what is meant by it

AnthonyLim23 replied:
could you ask Steve?

Matt-Cumber replied:
From Steve:
_The entry period_sequences records how many times the complete period sequence was executed during the run. For instance, if we look at EMU run 114295, which was a run carried out with a two period sequence collecting 500 frames in each period (i.e. 1000 frames to complete a single sequence), period_sequences reads 218 suggesting 218000 frames have been collected during the run. This compares well with frames_good (218002), a record of the total number of frames collected. (not sure why we’ve recorded two additional frames in the final period, but that’s another matter)._

AnthonyLim23 replied:
So in simple terms: a sequence is completing recording for all of the periods once?

Can we add a comment explaining this?

AnthonyLim23 commented on qt/widgets/common/test/MuonPeriodInfoTest.h:
```diff
@@ -0,0 +1,107 @@
+    TS_ASSERT_EQUALS(1, m_periodInfo->getDAQCount());
+    m_periodInfo->addPeriodToTable("state 1", "1", "10", "200", "25", "1");
+    m_periodInfo->addPeriodToTable("state 1", "1", "10", "200", "25", "1");
+    TS_ASSERT_EQUALS(3, m_periodInfo->getDAQCount());
```
what trigged the DAQ count to increase?

Matt-Cumber replied:
Adding a DAQ period to the table. the second argument is the type of period (1=DAQ, 2=DWELL). Here we added two daq periods as they both have the second argument as 1

AnthonyLim23 replied:
Could we have comment explaining that?

AnthonyLim23 commented on qt/widgets/common/test/MuonPeriodInfoTest.h:
```diff
@@ -0,0 +1,107 @@
+
+  void test_setNumberOfSequences() {
+    m_periodInfo->setNumberOfSequences(2);
+    TS_ASSERT_EQUALS("Run contains 2 cycles of periods", m_periodInfo->getNumberOfSequencesString());
```
does this mean that the sequences are the number of periods?

Matt-Cumber replied:
No they are not 2 is just a random arbitrary number I used for the test. This text format is the same as Wimbda

AnthonyLim23 replied:
ok. I just want to know what it means

Matt-Cumber committed:
Re #31798 Changes From Review

Added comment to explain DAQ count unit test. Added comment explaining number of sequences. Removed parent from Muon GUI's

AnthonyLim23 commented:
Opening FDA causes a crash

AnthonyLim23 commented:
When I open the periods window in MA or FDA, it sort of merges with the main Mantid tool bar. So I cannot move it.

AnthonyLim23 commented:
![image](https://user-images.githubusercontent.com/25006392/124737435-db332e00-df0f-11eb-9f83-1aec27b45fd0.png)

AnthonyLim23 reviewed (approved):
It all works now as expected. :shipit:
